File: src/main/java/ai/devchat/devchat/handler/CommitCodeRequestHandler.java
package ai.devchat.devchat.handler;

import ai.devchat.common.Log;
import ai.devchat.devchat.ActionHandler;
import ai.devchat.devchat.DevChatActionHandler;
import ai.devchat.devchat.DevChatActions;

import com.alibaba.fastjson.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.function.BiConsumer;

public class CommitCodeRequestHandler implements ActionHandler {
    private JSONObject metadata;
    private JSONObject payload;
    private final DevChatActionHandler devChatActionHandler;

    public CommitCodeRequestHandler(DevChatActionHandler devChatActionHandler) {
        this.devChatActionHandler = devChatActionHandler;
    }

    @Override
    public void executeAction() {
        Log.info("Handling commit code request");

        String callbackFunc = metadata.getString("callback");
        String message = payload.getString("message");
        String[] commitCommand = {"git", "commit", "-m", message};

        StringBuilder result = new StringBuilder();
        StringBuilder error = new StringBuilder();

        BufferedReader reader = null;
        BufferedReader errorReader = null;

        try {
            String projectDir = devChatActionHandler.getProject().getBasePath();
            Log.info("Preparing to execute command: git commit -m " + message + " in " + projectDir);
            Process process = Runtime.getRuntime().exec(commitCommand, null, new File(projectDir));
            reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                result.append(line).append("\n");
            }

            errorReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
            while ((line = errorReader.readLine()) != null) {
                error.append(line).append("\n");
            }
        } catch (IOException e) {
            error.append(e.getMessage());
        } finally {
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (errorReader != null) {
                try {
                    errorReader.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        BiConsumer<JSONObject, JSONObject> processCommitResponse = (metadata, payload) -> {
            if (error.isEmpty()) {
                metadata.put("status", "success");
                metadata.put("error", "");
            } else {
                metadata.put("status", "error");
                metadata.put("error", error.toString());
            }
        };

        devChatActionHandler.sendResponse(DevChatActions.COMMIT_CODE_RESPONSE, callbackFunc, processCommitResponse);
    }

    @Override
    public void setMetadata(JSONObject metadata) {
        this.metadata = metadata;
    }

    @Override
    public void setPayload(JSONObject payload) {
        this.payload = payload;
    }
}

File: src/main/kotlin/ai/devchat/devchat/handler/CommitCodeRequestHandler.kt
package ai.devchat.devchat.handler

import ai.devchat.common.ProjectUtils
import ai.devchat.common.Log
import ai.devchat.devchat.BaseActionHandler
import ai.devchat.devchat.DevChatActions
import com.alibaba.fastjson.JSONObject
import java.io.BufferedReader
import java.io.File

class CommitCodeRequestHandler(requestAction: String, metadata: JSONObject?, payload: JSONObject?) : BaseActionHandler(
    requestAction,
    metadata,
    payload
) {
    override val actionName: String = DevChatActions.COMMIT_CODE_RESPONSE

    override fun action() {
        val message = payload!!.getString("message")
        val command = arrayOf("git", "commit", "-m", message)
        val projectDir = ProjectUtils.project?.basePath
        val process = Runtime.getRuntime().exec(command, null, projectDir?.let { File(it) })
        val result = process.inputStream.bufferedReader().use(BufferedReader::readText)
        val errors = process.errorStream.bufferedReader().use(BufferedReader::readText)
        process.waitFor()
        val exitCode = process.exitValue()
        if (exitCode != 0) {
            throw RuntimeException("Failed to execute command: $command, Exit Code: $exitCode Error: $errors")
        }
        Log.info(result)
        super.action()
    }
}